### PR TITLE
Bluetooth

### PR DESCRIPTION
**Pull request by droidizer**
Removed speakerphone settings and Enabled Bluetooth by default.

In case there is no bluetooth it defaults to the system settings.

### CODE REVIEW

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java [hunk 1]
```diff
@@ -75,6 +75,9 @@ public void receiveCommand(CustomTwilioVideoView view, int commandId, @Nullable
                 boolean enableAudio = args.getBoolean(2);
                 boolean enableVideo = args.getBoolean(3);
                 boolean enableRemoteAudio = args.getBoolean(4);
+
+                //enable bluetooth by default
+                view.toggleBluetoothHeadset(true);
```
Comment (a-morenovalls) [inline android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: In my opinion I would remove this line and configure bluetooth on setAudioFocus method. Then, call `stopBluetoothSco()` when the component or activity is destroyed (onHostDestroy lifecycle method).

Comment (droidizer) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: This method was already existing to handle bluetooth config so I didn't call the stop. But makes sense.

**Comment by DeTeam** [comment]
@droidizer could you share some info on how audio routing works here and which cases are not covered by this approach (connecting Bluetooth headsets in the middle of the call)?

**Comment by DeTeam** [comment]
Seems like for nice audio output device switching twilio video quickstart guide (https://github.com/twilio/video-quickstart-android/blob/master/quickstart/src/main/java/com/twilio/video/quickstart/activity/VideoActivity.java#L210) for android is using https://github.com/twilio/audioswitch

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java (hunk 1 above)
Comment (a-morenovalls) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: Yes, but makes more sense to remove all bluetooth or methods we don't rely on more.

**Comment by a-morenovalls** [comment]
> Seems like for nice audio output device switching twilio video quickstart guide (https://github.com/twilio/video-quickstart-android/blob/master/quickstart/src/main/java/com/twilio/video/quickstart/activity/VideoActivity.java#L210) for android is using https://github.com/twilio/audioswitch

As far as I can see, the library automatically handles the connection and prioritises bluetooth connection:
https://github.com/twilio/audioswitch#select-a-device

Probably can be a good idea to try this :)

**Comment by DeTeam** [comment]
Since we don't give an option to choose devices in the app we can also build a hierarchy for output devices:

"wired, Bluetooth, loudspeaker", sort  and pick the first available + do the same whenever something changes

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java (hunk 1 above)
Comment (DmitryTop) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: BTW. `toggleBluetoothHeadset` can be call from js:  `TwilioVideoRef.setBluetoothHeadsetConnected(true)`. There is no reason to change the plugin.

Comment (a-morenovalls) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: Yes, but I think @DeTeam commented that he wanted to change the plugin instead calling these methods on the JS.

**Comment by DeTeam** [comment]
https://github.com/twilio/audioswitch/issues/45#issuecomment-671999638 + other comments with some ideas on the implementation

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java [hunk 2]
```diff
@@ -450,7 +456,7 @@ private void setAudioFocus(boolean focus) {
              * speaker mode if this is not set.
              */
             audioManager.setMode(AudioManager.MODE_IN_COMMUNICATION);
-            audioManager.setSpeakerphoneOn(!audioManager.isWiredHeadsetOn());
+            audioManager.setSpeakerphoneOn(false);
```
Comment (DmitryTop) [inline android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java:459]: What if wired headset is not connected?

Comment (droidizer) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java:459]: Bluetooth needs speakerPhone Off settings to function and if it cant find a bluetooth the system defaults to speakerphone anyway.

**Comment by droidizer** [comment]
1. At any given time the Bluetooth headset is going to be used if there is one available with the microphone. If not it just uses headset for audio streaming. 
2. Speakerphone settings are needed to be off mandatorily for bluetooth to work. 
3. If Bluetooth gets disconnected during a call again system default speaker and microphone are used. 
4. If another bluetooth gets connected when the class is running the bluetooth headset disconnects, currently the class cannot establish a connection to the latest device.
To connect to the latest bluetooth headset the class has to be restarted as the bluetooth and audio connections are configured on Connecting To Room. 

So basically, 
1. On start of a Class/Session if bluetooth headset is connected then the headset is used. 
2. If the Bluetooth is connected later after starting the session the connection cannot be established. 
3. For handling the case where the app needs to support multiple bluetooth headsets. Then audioswitch library can be used as it posts connection events. But the selection of device will have to be handled by the app. 
Another way is to have your own listener to connection states: But the code is a bit outdated. 
https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/audio/BtHelper.java
4. In case bluetooth connection toggles, the session cannot reconnect to bluetooth so the session needs to be restarted. 

Hope this helps. Please let me know in case of any other questions. 

@DeTeam I tried the audiodswitch lib approach first but it didn't work as the webrtc plugin was overriding my changes. So decided to move the code here. I think it makes sense to make these changes after Audio is acquired if you need to use audioswitch then webrtc plugin needs to have no logic consisting of AudioFocus. So you can use apprtc.

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java (hunk 1 above)
Comment (droidizer) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoViewManager.java]: @Moreno97 I haven't removed logic  of TOGGLE BLUETOOTH HEADSET configuration if RN wants to call this in the future you can. Except the command doesn't get posted for some reason, setBluetoothHeadsetConnected(true) on TwilioVideoView didn't work for me so needs investigation.

**Comment by DeTeam** [comment]
Tested on oneplus nord, android 10, bose sport headphones, didn't work (connected headphones before the session, check the sound via spotify).

**Pull request merged**